### PR TITLE
Add log throttling

### PR DESCRIPTION
**Pull request by freiksenet**

### CODE REVIEW

File: python_modules/dagit/dagit/pipeline_run_storage.py [hunk 1]
```diff
@@ -50,6 +52,25 @@ def __init__(self, run_id, pipeline_name, config, execution_plan):
...
+            self._queue_timeout = time.time()
+            self._flush_queued = True
+            while (time.time() - self._queue_timeout) < 1:
+                gevent.sleep(0.1)
+            with self._log_queue_lock:
+                if self._log_queue:
+                    events = self._log_queue
+                    self._log_queue = []
+
+                    for subscriber in self._subscribers:
```
Comment (schrockn) [inline python_modules/dagit/dagit/pipeline_run_storage.py]: it's a bit scary to me to call into all the rx machinery etc while holding a lock. why don't we copy the log queue into the events array and then release the lock?

File: python_modules/dagit/dagit/pipeline_run_storage.py [hunk 2]
```diff
@@ -50,6 +52,25 @@ def __init__(self, run_id, pipeline_name, config, execution_plan):
...
+        self._log_queue = []
+        self._flush_queued = False
+        self._queue_timeout = None
+
+    def _enqueue_flush_logs(self):
+        if not self._flush_queued:
+            self._queue_timeout = time.time()
+            self._flush_queued = True
+            while (time.time() - self._queue_timeout) < 1:
+                gevent.sleep(0.1)
```
Comment (schrockn) [inline python_modules/dagit/dagit/pipeline_run_storage.py]: This while loop is worthy of a comment re: why 1 and .1 was chosen and what is going on here

File: python_modules/dagit/dagit/pipeline_run_storage.py [hunk 3]
```diff
@@ -50,6 +52,29 @@ def __init__(self, run_id, pipeline_name, config, execution_plan):
         self.pipeline_name = pipeline_name
         self.config = config
         self.execution_plan = execution_plan
+        self._log_queue_lock = gevent.lock.Semaphore()
+        self._log_queue = []
+        self._flush_queued = False
+        self._queue_timeout = None
+
+    def _enqueue_flush_logs(self):
+        if not self._flush_queued:
```
Comment (schrockn) [inline python_modules/dagit/dagit/pipeline_run_storage.py]: let's put _flushed_queued behind a lock. in particular I think we need to test for false and set to true atomically

File: python_modules/dagit/dagit/pipeline_run_storage.py [hunk 4]
```diff
@@ -50,6 +54,35 @@ def __init__(self, run_id, pipeline_name, config, execution_plan):
...
         self.execution_plan = execution_plan
+        self._log_queue_lock = gevent.lock.Semaphore()
+        self._log_queue = []
+        self._flush_queued = False
+        self._queue_timeout = None
+
+    def _enqueue_flush_logs(self):
+        with self._log_queue_lock:
+            if not self._flush_queued:
+                self._queue_timeout = time.time()
```
Comment (schrockn) [inline python_modules/dagit/dagit/pipeline_run_storage.py]: we could move this outside the lock scope (the time.time() call)

Comment (freiksenet) [reply python_modules/dagit/dagit/pipeline_run_storage.py]: I think it's fine there, just so we couldn't override it by accident if logic changes at some point.

**Pull request merged**